feature: keyboard - encoding using differen layouts e.g. QWERTY, QWERTZ, Dvorak, RISTOME, COLEMAK

t-m-z opened the pull request:
HELP - There is one problem: switchMapKeyValue forgets to switch some keys - e.g. switching Dvorak

S-Man42 commented on assets/i18n/de.json:
```diff
@@ -1248,6 +1248,21 @@
+  "keyboard_title" : "Tastatur",
+  "keyboard_description" : "Konvertiert zwischen Tastaturlayouts QWERTZ, QWERTY, DVORAK, COLEMAK, RISTOME, NEO",
+  "keyboard_example" : "QWERTZ → Ü;:PY",
+  "keyboard_from" : "Konvertiert von",
```
I'd simply write "from/to" or "von/nach". This is more convenient to the current app

S-Man42 commented on assets/i18n/de.json:
```diff
@@ -1248,6 +1248,21 @@
   "kenny_case_sensitive" : "Beachte Groß-/Kleinschreibung",
 
+  "keyboard_title" : "Tastatur",
+  "keyboard_description" : "Konvertiert zwischen Tastaturlayouts QWERTZ, QWERTY, DVORAK, COLEMAK, RISTOME, NEO",
```
Could you please add the French AZERTY layout too? :)

S-Man42 commented on lib/widgets/registry.dart:
```diff
@@ -440,6 +441,12 @@ class Registry {
+      GCWTool(
+        tool: Keyboard(),
+        i18nPrefix: 'keyboard',
+        category: ToolCategory.CRYPTOGRAPHY,
```
For me, it's clear SCIENCE_AND_TECHNOLOGY... Especially technology. This is where phone keys is located too

S-Man42 commented on lib/widgets/registry.dart:
```diff
@@ -440,6 +441,12 @@ class Registry {
+        tool: Keyboard(),
+        i18nPrefix: 'keyboard',
+        category: ToolCategory.CRYPTOGRAPHY,
+        searchStrings: 'dvorak keyboard tastatur layout qwertz qwerty neo colemak ristome'
```
maybe add: "keys tasten"

S-Man42 commented on lib/widgets/tools/crypto_and_encodings/keyboard.dart:
```diff
@@ -0,0 +1,143 @@
+            });
+          },
+        ),
+        Row(
```
I'd write 1 Row containing 2 Columns. With that, you are able to add a little space between both columns (Pseudocode,  I am not quite sure about correct names, but this could be the schema):

Row(
  children: [
     Container(
       child: Column () // add Divider and DropDown,
      padding: EdgeInsectOnly(right: DEFAULT_MARGIN)
     ),
    Container(
        child: Column () // other Divider and Dropdown,
        padding: EdgeInsectsOnly(left: DEFAULT_MARGIN)
    )
  ]
)

S-Man42 reviewed (commented):
- Converting from DEU T1 to US int: Ü -> returns Ü. I guess, thats incorrect. Maybe there are other problems, please check the maps once more :)

Code is fine! Great, thanks!

S-Man42 commented:
http://fbcs.bplaced.net/multi_encoder_decoder.html -> DVORAK I, DVORAK II US, DVORAK II DE

t-m-z commented:
Refactor Frontend - changing rows and colums
Moving from Codes... to Science...
Adding two maps for FRA AZERTY, BEPO
Adding tests
checking maps

S-Man42 commented on lib/widgets/registry.dart:
```diff
@@ -92,6 +92,7 @@ import 'package:gc_wizard/widgets/tools/crypto_and_encodings/chao.dart';
 import 'package:gc_wizard/widgets/tools/crypto_and_encodings/chicken_language.dart';
 import 'package:gc_wizard/widgets/tools/crypto_and_encodings/deadfish.dart';
 import 'package:gc_wizard/widgets/tools/crypto_and_encodings/duck_speak.dart';
+import 'file:///D:/Programmierung/GCWizard/lib/widgets/tools/science_and_technology/keyboard.dart';
```
You have to fix the absolute path

S-Man42 commented on lib/widgets/main_view.dart:
```diff
@@ -52,6 +52,7 @@ import 'package:gc_wizard/widgets/tools/crypto_and_encodings/chao.dart';
 import 'package:gc_wizard/widgets/tools/crypto_and_encodings/chicken_language.dart';
 import 'package:gc_wizard/widgets/tools/crypto_and_encodings/deadfish.dart';
 import 'package:gc_wizard/widgets/tools/crypto_and_encodings/duck_speak.dart';
+import 'file:///D:/Programmierung/GCWizard/lib/widgets/tools/science_and_technology/keyboard.dart';
```
You have to fix the absolute path here as well

S-Man42 commented on lib/widgets/tools/science_and_technology/keyboard.dart:
```diff
@@ -0,0 +1,115 @@
+import 'package:flutter/material.dart';
+import 'package:gc_wizard/i18n/app_localizations.dart';
+import 'file:///D:/Programmierung/GCWizard/lib/logic/tools/science_and_technology/keyboard.dart';
```
absolut path... This happens when you move a already referenced file... It's bad logic of the IDE intelligence

S-Man42 commented on test/logic/tools/science_and_technology/keyboard.dart:
```diff
@@ -0,0 +1,309 @@
+import "package:flutter_test/flutter_test.dart";
+import 'file:///D:/Programmierung/GCWizard/lib/logic/tools/science_and_technology/keyboard.dart';
```
...

S-Man42 commented on lib/logic/tools/science_and_technology/keyboard.dart:
```diff
@@ -0,0 +1,226 @@
+}
+
+String encodeKeyboard(String input, KeyboardLayout keyboardFrom, keyboardTo){
+print(keyboardFrom.toString()+'=>'+keyboardTo.toString());
```
Please remove the print() call

S-Man42 commented on lib/widgets/tools/science_and_technology/keyboard.dart:
```diff
@@ -0,0 +1,115 @@
+        ),
+        Row(
+          children: <Widget>[
+            Expanded(
```
In the previous review I mentioned to add some space between the DropDowns. Please add this :)

S-Man42 commented on lib/logic/tools/science_and_technology/keyboard.dart:
```diff
@@ -0,0 +1,226 @@
+    KeyboardLayout.Dvorak_I_DEU3 : i18n(context, 'keyboard_mode_dvorak_I3'),
+    KeyboardLayout.Dvorak_II_DEU : i18n(context, 'keyboard_mode_dvorak_II'),
+    KeyboardLayout.FRA_AZERTY : i18n(context, 'keyboard_mode_fra_azerty'),
+    KeyboardLayout.FRA_BEPO : i18n(context, 'keyboard_mode_fra_bepo'),
```
The i18n string is named "keyboard_mode_bebo" without "fra"

S-Man42 commented on lib/logic/tools/science_and_technology/keyboard.dart:
```diff
@@ -0,0 +1,226 @@
+
+Map buildKeyboardsMap(var context) {
+  return {
+    KeyboardLayout.QWERTZ_T1 : i18n(context, 'keyboard_mode_qwertz_t1'),
```
Please don't move the i18n() call into the logic. That's a pure frontend function. Please call this in the GCWDropDownItem-> child: i18n(context, mode.value)

S-Man42 commented on assets/i18n/en.json:
```diff
@@ -1248,6 +1248,24 @@
+  "keyboard_mode_qwertz_t1" : "QWERTZ DEU T1",
+  "keyboard_mode_qwerty_us_int" : "QWERTY US int",
+  "keyboard_mode_dvorak" : "DVORAK",
+  "keyboard_mode_dvorak_I1" : "DVORAK Typ I.1 German Ä,.PYF",
```
Maybe you could split this into "DVORAK Typ I.1 German" and as subtitle (see below): "Ä,.PYF"

S-Man42 commented on assets/i18n/en.json:
```diff
@@ -1248,6 +1248,24 @@
+  "keyboard_mode_dvorak_I1" : "DVORAK Typ I.1 German Ä,.PYF",
+  "keyboard_mode_dvorak_I2" : "DVORAK Typ I.2 German ÖÜÄPYF",
+  "keyboard_mode_dvorak_I3" : "DVORAK Typ I.3 German ÄÖÜPYF",
+  "keyboard_mode_dvorak_II" : "DVORAK Typ II German",
```
In English: "Type"

S-Man42 commented on lib/widgets/tools/science_and_technology/keyboard.dart:
```diff
@@ -0,0 +1,115 @@
+                      });
+                    },
+                    items: KeyboardModeItemsTo.entries.map((mode) {
+                      return GCWDropDownMenuItem(
```
You could use the new "subtitle" attribute to show the example (mentioned above):
```
return GCWDropDownMenuItem(
              value: ...,
              child: ...,
              subtitle: mode.example
            )
```
A good example can be found at gcw_coords_formatselector.dart

S-Man42 commented on lib/widgets/tools/science_and_technology/keyboard.dart:
```diff
@@ -0,0 +1,115 @@
+            Expanded(
+              flex: 1,
+              child: Column(
+                children: <Widget>[
```
Space between Dropdowns:

https://github.com/S-Man42/GCWizard/pull/260#discussion_r534310537

t-m-z commented:
test keyboard.dart
- absolute path => removed and changed to relative path

logic keyboard.dart
- print call => removed
- i18n(context, ... => removed

mainview
- absolute path => removed and changed to relative path

en.json, de.json
- keyboard_mode_bepo => keyboard_mode_fra_bepo

en.json
- Typ => Type

widget keyboard.dart
- absolute path => removed and changed to relative path
- subtitle => added
- padding =>  padding: EdgeInsets.only(right: DEFAULT_MARGIN)
